### PR TITLE
Share karma among co-authors

### PR DESCRIPTION
**Pull request by Will-Howard**
For posts with co-authors, give each author karma/sqrt(numAuthors) each.

### CODE REVIEW

File: packages/lesswrong/server/callbacks/votingCallbacks.ts [hunk 1]
```diff
@@ -4,44 +4,46 @@ import { voteCallbacks, VoteDocTuple } from '../../lib/voting/vote';
 import { postPublishedCallback } from '../notificationCallbacks';
 import { batchUpdateScore } from '../updateScores';
```
Comment (Will-Howard) [inline packages/lesswrong/server/callbacks/votingCallbacks.ts:6]: You might want to look at this file in side-by-side view as I have refactored a bit. It shouldn't have changed the behaviour at all in the one user case though

File: packages/lesswrong/server/callbacks/votingCallbacks.ts [hunk 2]
```diff
@@ -4,44 +4,46 @@ import { voteCallbacks, VoteDocTuple } from '../../lib/voting/vote';
...
+voteCallbacks.castVoteAsync.add(({newDocument, vote}: VoteDocTuple) => applyKarmaToUsers(usersVoteAppliesTo(newDocument, vote), vote.power));
+voteCallbacks.cancelAsync.add(({newDocument, vote}: VoteDocTuple) => applyKarmaToUsers(usersVoteAppliesTo(newDocument, vote), -vote.power));
 
-  if (newDocument.userId !== vote.userId) {
-    void Users.rawUpdateOne({_id: vote.userId}, {$inc: {[field]: -1, voteCount: -1}});
-  }
-});
+// update user vote counts
+// NOTE: currently this has the slightly confusing behaviour that if you upvote a post, and then cancel that upvote, your overall vote count is increased by 1
+// (the initial upvote gives +1, the cancelled upvote gives -1, and a "neutral" vote is added with +1)
```
Comment (Will-Howard) [inline packages/lesswrong/server/callbacks/votingCallbacks.ts:44]: I have just left this behaviour as is, whether this is a bug or not depends how you interpret "vote count" 🤔

File: packages/lesswrong/server/callbacks/votingCallbacks.ts [hunk 3]
```diff
@@ -4,44 +4,46 @@ import { voteCallbacks, VoteDocTuple } from '../../lib/voting/vote';
...
 
-  if (newDocument.userId !== vote.userId) {
-    void Users.rawUpdateOne({_id: vote.userId}, {$inc: {[field]: 1, voteCount: 1}});
-  }
-});
+  const field = vote.voteType + "Count";
+  void Users.rawUpdateOne({_id: vote.userId}, {$inc: {[field]: increment, voteCount: increment}});
+}
 
-voteCallbacks.cancelAsync.add(async function cancelVoteCount ({newDocument, vote}: VoteDocTuple) {
```
Comment (Will-Howard) [inline packages/lesswrong/server/callbacks/votingCallbacks.ts:38]: these functions used to be async but I don't think they needed to be, correct me if I'm wrong

**Review by jpaddison3 — APPROVED** [review]

**Comment by Discordius** [comment]
Alas, we'll likely have to revert this PR. It is quite important that the vote totals are directly derivable from nothing but the vote table, so that we can rerun vote totals from time to time, so any change to how karma gets allocated also needs to change the code for rerunning vote totals.

**Comment by Discordius** [comment]
@jpaddison3

**Comment by jpaddison3** [comment]
My apologies for forgetting about the places where we rederive karma. We searched for places where karma was set, but we missed it. I should nevertheless have remembered — I know I've seen it in the merge user script at least.

However it seems like a pretty easy remedy? Find the places where we recompute karma and fix those places to take this into account. I can see a few complications, but it seems basically fine to me. I will make sure we do this.

In the mean time this can continue running. The situation with this PR seems better than the situation without it to me.

**Pull request merged**